internal/apex/ctlserver.go

```diff
@@ -0,0 +1,119 @@
+package apex
+
+import (
+	"fmt"
+	"net"
+	"net/rpc"
+	"net/rpc/jsonrpc"
+	"os"
+	"sync"
+)
+
+// TODO make this path configurable
+const UnixSocketPath = "/run/apex.sock"
+
+func (ax *Apex) CtlServerStart() error {
+	switch ax.os {
+	case Linux.String():
+		return ax.CtlServerStartLinux()
+
+	case Darwin.String():
+		ax.logger.Info("Ctl interface not yet supported on OSX")
+
+	case Windows.String():
+		ax.logger.Info("Ctl interface not yet supported on Windows")
+	}
+
+	return nil
+}
+
+type ApexCtl struct {
+	ax *Apex
+}
+
+func (ac *ApexCtl) Status(_ string, result *string) error {
+	var statusStr string
+	switch ac.ax.status {
+	case ApexStatusStarting:
+		statusStr = "Starting"
+	case ApexStatusAuth:
+		statusStr = "WaitingForAuth"
+	case ApexStatusRunning:
+		statusStr = "Running"
+	default:
+		statusStr = "Unknown"
+	}
+	res := fmt.Sprintf("Status: %s\n", statusStr)
+	if len(ac.ax.statusMsg) > 0 {
+		res += ac.ax.statusMsg
+	}
+	*result = res
+	return nil
+}
+
+func (ac *ApexCtl) Version(_ string, result *string) error {
+	*result = ac.ax.version
+	return nil
+}
+
+func (ax *Apex) CtlServerStartLinux() error {
+	os.Remove(UnixSocketPath)
+	l, err := net.ListenUnix("unix", &net.UnixAddr{Name: UnixSocketPath, Net: "unix"})
+	if err != nil {
+		ax.logger.Error("Error creating unix socket: ", err)
+		return err
+	}
+	ax.wg.Add(1)
+	go func() {
+		defer l.Close()
+		defer ax.wg.Done()
+
+		ac := new(ApexCtl)
+		ac.ax = ax
+		err := rpc.Register(ac)
+		if err != nil {
+			ax.logger.Errorln("Error on rpc.Register(): ", err)
```

Comment:
This seems like a fatal error right? We should be returning here and not just logging

Reply:
Oops, yes.